bsp/src/org/jetbrains/bsp/project/test/BspTestRunProfileState.scala

```diff
@@ -0,0 +1,145 @@
+package org.jetbrains.bsp.project.test
+
+import java.io.OutputStream
+import java.net.URI
+import java.util.UUID
+import java.util.concurrent.CompletableFuture
+
+import ch.epfl.scala.bsp4j.TaskDataKind._
+import ch.epfl.scala.bsp4j.TestStatus._
+import ch.epfl.scala.bsp4j._
+import com.google.gson.{Gson, JsonObject}
+import com.intellij.execution.configurations.{RunProfileState, RunnerSettings}
+import com.intellij.execution.process.{ProcessHandler, ProcessOutputType}
+import com.intellij.execution.runners.ProgramRunner
+import com.intellij.execution.testframework.sm.SMTestRunnerConnectionUtil
+import com.intellij.execution.testframework.sm.runner.SMTRunnerConsoleProperties
+import com.intellij.execution.{DefaultExecutionResult, ExecutionResult, Executor}
+import com.intellij.openapi.module.ModuleManager
+import com.intellij.openapi.project.Project
+import jetbrains.buildServer.messages.serviceMessages._
+import org.jetbrains.bsp.data.BspMetadata
+import org.jetbrains.bsp.protocol.BspCommunicationComponent
+import org.jetbrains.bsp.protocol.BspNotifications.{BspNotification, TaskFinish, TaskStart}
+import org.jetbrains.bsp.protocol.session.BspSession.BspServer
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.concurrent.ExecutionContext.Implicits._
+
+class BspTestRunProfileState(val project: Project, val rc: BspTestRunConfiguration, val ex: Executor) extends RunProfileState {
+
+  val gson = new Gson()
+
+  class MProcHandler extends ProcessHandler {
+    override def destroyProcessImpl(): Unit = ()
+
+    override def detachProcessImpl(): Unit = {}
+
+    override def detachIsDefault(): Boolean = false
+
+    override def getProcessInput: OutputStream = null
+
+    def shutdown(): Unit = {
+      super.notifyProcessTerminated(0)
+    }
+  }
+
+  private def targets(): List[URI] = {
+    ModuleManager.getInstance(project).getModules.toList
+      .flatMap(BspMetadata.get(project, _))
+      .flatMap(x => x.targetIds.asScala.toList)
+  }
+
+  private def testRequest(server: BspServer): CompletableFuture[TestResult] = {
+    val targetIds = targets().map(uri => new BuildTargetIdentifier(uri.toString))
+    val params = new TestParams(targetIds.toList.asJava)
+    params.setOriginId(UUID.randomUUID().toString)
+    server.buildTargetTest(params)
+  }
+
+
+  def printProc(str: MessageWithAttributes)(implicit proc: ProcessHandler): Unit = {
+    proc.notifyTextAvailable(str + System.lineSeparator(), ProcessOutputType.STDOUT)
+  }
+
+  class ServiceMsg(val name: String, val map: Map[String, String]) extends MessageWithAttributes(name, map.asJava) {
+    def this(n: String) = this(n, Map())
+  }
+
+  case class NestedBspMsg(nested: AnyRef, msg: AnyRef)
+
+  type HasNested = {
+    def getDataKind: String
+    def getData: Object
+  }
+
+  private def extractNestedMessage(bspMessage: AnyRef): AnyRef = bspMessage match {
+    case x: HasNested@unchecked =>
```

Comment:
Used reflection here, to avoid repeating code near match expressions. Maybe it's not a very good idea.

Reply:
Looks okay, but you might want to look into the `unapply` mechanism for a pattern matching approach that doesn't require type reflection